core/src/View/Hooks/Kokkos_ViewHooks.hpp

```diff
@@ -0,0 +1,188 @@
+struct invoke_subscriber_impl<Invoker, Subscriber, RemSubscribers...> {
+  template <typename ViewType>
+  static void invoke(ViewType &self, const ViewType &other) {
+    Invoker<Subscriber>::call(self, other);
+    invoke_subscriber_impl<Invoker, RemSubscribers...>::invoke(self, other);
+  }
+};
+
+template <typename Subscriber>
+struct copy_constructor_invoker {
+  template <typename View>
+  static void call(View &self, const View &other) {
+    Subscriber::copy_constructed(self, other);
+  }
+};
+
+template <typename Subscriber>
+struct move_constructor_invoker {
+  template <typename View>
+  static void call(View &self, const View &other) {
+    Subscriber::move_constructed(self, other);
+  }
+};
+
+template <typename Subscriber>
+struct copy_assign_invoker {
+  template <typename View>
+  static void call(View &self, const View &other) {
+    Subscriber::copy_assigned(self, other);
+  }
+};
+
+template <typename Subscriber>
+struct move_assign_invoker {
+  template <typename View>
+  static void call(View &self, const View &other) {
+    Subscriber::move_assigned(self, other);
+  }
+};
+}  // namespace Detail
+
+struct EmptyViewHooks {
+  using hooks_policy = EmptyViewHooks;
+
+  template <typename View>
+  static void copy_construct(View &, const View &) {}
+  template <typename View>
+  static void copy_assign(View &, const View &) {}
+  template <typename View>
+  static void move_construct(View &, const View &) {}
+  template <typename View>
+  static void move_assign(View &, const View &) {}
+};
+
+template <class... Subscribers>
+struct SubscribableViewHooks {
+  using hooks_policy = SubscribableViewHooks<Subscribers...>;
+
+  template <typename View>
+  static void copy_construct(View &self, const View &other) {
+    Detail::invoke_subscriber_impl<Detail::copy_constructor_invoker,
+                                   Subscribers...>::invoke(self, other);
+  }
+  template <typename View>
+  static void copy_assign(View &self, const View &other) {
+    Detail::invoke_subscriber_impl<Detail::copy_assign_invoker,
+                                   Subscribers...>::invoke(self, other);
+  }
+  template <typename View>
+  static void move_construct(View &self, const View &other) {
+    Detail::invoke_subscriber_impl<Detail::move_constructor_invoker,
+                                   Subscribers...>::invoke(self, other);
+  }
+  template <typename View>
+  static void move_assign(View &self, const View &other) {
+    Detail::invoke_subscriber_impl<Detail::move_assign_invoker,
+                                   Subscribers...>::invoke(self, other);
+  }
+};
+
+namespace Impl {
+template <class ViewType, class Traits = typename ViewType::traits,
+          class Enabled = void>
+struct DynamicViewHooksCaller;
+}
+
+struct DynamicViewHooksSubscriber {
+  template <typename View>
+  static void copy_constructed(View &self, const View &) {
+    Impl::DynamicViewHooksCaller<View>::call_copy_construct_hooks(self);
```

Comment:
Why do the dynamic hooks only take one argument and ignore the other?

Reply:
I was back and forth on this. Dynamic view hooks don't need it but maybe another application might. It's probably not worth it to provide the second argument, so I'll see about changing that